### DISCUSSION TITLE
Refetch remote emojis on accounts name and bio

### DISCUSSION THREAD

**Discussion by emanuelpina** [mastodon/mastodon#27923] [Q&A]
On an atempt to delete all the local custom emojis on my instance I used tootctl command:

```
bin/tootctl emoji purge --no-remote-only
```

Despite using the flag `--no-remote-only` the remote custom emojis were deleted too.

No big fuss, Mastodon just start fetching the remote emojis again. The problem is this is only true for emojis on toots. For accounts that use custom emojis on the Display name and Bio, Mastodon isn’t refetching the emojis for the already known accounts. For new discovered accounts it fetch the emojis as usual.

I tried using tootctl to refresh those remote accounts, without success.

There’s any way I can force the refetch of those emojis?

**Comment by ChlorideCull** [comment]
They should get pulled in when you get activity from that user, and it's been more than a day since it last updated, which triggers a webfinger and subsequent account sync.

The issue is two-fold in this case:

1. There's no way to just remove local custom emoji with tootctl.
2. `tootctl accounts refresh` does **not** fetch remote user data as it claims, all it does is redownload the avatar and header. I'll be making a PR that changes it to use `ActivityPub::FetchRemoteAccountService` instead to properly refresh everything to do with the accounts, unless someone else gets around to it.

**Comment by teutat3s** [comment]
Came across this today, too - would be glad to find a workaround.

**Comment by Denommus** [comment]
I also had this come up today, it's kinda annoying.

**Comment by keskival** [comment]
This is still annoying after 2 years with no real workaround.

**Comment by keskival** [comment]
There is some related information in these two places, but the instructions are not very clear or up-to-date:
- https://discourse.joinmastodon.org/t/can-the-cache-directory-be-removed-from-backups/3683/2
- https://discourse.joinmastodon.org/t/restored-backup-without-cache-dir-manual-removal-did-not-help/3772
This also seems to relate, but while it was approved, it was never merged to Mastodon, but to Hometown Fork :(, Also, it's about media attachments, not about emojis, but the problem space is intertwined in the "I didn't think I needed to back up the caches" space:
- https://github.com/mastodon/mastodon/pull/16641

There are several scripts executable with `RAILS_ENV=production bundle exec rails c`, for convenience laid out here with minor corrections. Do not execute these without understanding what they do!:

Deleting all cached media attachments from external servers to allow web app to refetch them when they are queried. At least on my instance this is reaaally slow, takes tens of seconds per attachment (is this the same as `tootctl media remove`?):
```
MediaAttachment.cached.where.not(remote_url: '').each do |attachment|
  attachment.file.destroy
  attachment.thumbnail.destroy
  attachment.save
end
```

Deleting and redownloading all cached remote emojis. This apparently has a problem that the local URLs for the cached emojis change, and the URLs aren't updated to the fetched profile headers. I suppose this should be done before the next steps, assuming the cached local emoji URLs are set up correctly to the profile headers and names at the time of fetching. This doesn't seem to work though, maybe `tootctl emoji purge --remote-only` would be better:
```
CustomEmoji.remote.where.not(image_file_name: nil).where.not(image_remote_url: '').each do |emoji|
  emoji.image.destroy
  emoji.image.save
  emoji.download_image!
end
```

Deleting all remote profile avatars, and refetching them. Are these two the same as `tootctl accounts refresh --all`?:
```
Account.remote.where.not(avatar_file_name: nil).where.not(avatar_remote_url: '').each do |a|
  a.avatar.destroy
  a.save
  a.download_avatar!
end
```

Deleting all remote profile headers, and refetching them. Hopefully this uses/updates the newly fetched local cached emoji URLs, and also includes the profile name as they often contain emojis:
```
Account.remote.where.not(header_file_name: nil).where.not(header_remote_url: '').each do |a|
  a.header.destroy
  a.save
  a.download_header!
end
```

It would be awesome to get tooling for these to `tootctl`. That said, I'm not entirely sure if destroying the entities is necessary before downloading them again, and perhaps we could use `remove_entity_cache` instead. I'm not very well acquainted with the codebase.

**Comment by zeroesca** [comment]
Just ran into this after an instance migrate + cloud storage provider change + proxy domain change. URLs of custom emojis in usernames/headers of remote accounts point to correct domain, but non-existing filesnames in /cache/../../.. -- even though in /admin/custom_emojis there are working copies. Refreshing the profile from the admin UI doesn't re-cache them.

**Comment by RokkuCode** [comment]
So the issue has been converted to a discussion. how is the plan on fixing this now? the problem still exists in 2024
